src/Components/Popover/Popover.tsx

```diff
@@ -0,0 +1,476 @@
+      acceptablePopoverPositions.push(PopoverPosition.ToTheRight)
+    }
+
+    // Check to see if the specified position is within the acceptable positions
+    if (acceptablePopoverPositions.includes(position)) {
+      return position
+    }
+
+    // Otherwise choose the next available position in from the acceptable list
+    return acceptablePopoverPositions[0] || null
+  }
+
+  private isDialogFlush = (position: PopoverPosition): PopoverFlush => {
+    const {triggerRect} = this.state
+
+    if (!triggerRect || !this.dialogRef.current) {
+      return {
+        first: false,
+        last: false,
+      }
+    }
+
+    const dialogRect = this.dialogRef.current.getBoundingClientRect()
+    // When the trigger is in a corner of the screen,
+    // determine whether to offest it
+    let first = false
+    let last = false
+
+    switch (position) {
+      case PopoverPosition.Above:
+      case PopoverPosition.Below:
+        // When the dialog is above or below the trigger
+        // First: left edge
+        // Last: right edge
+        const overflowX = dialogRect.width - triggerRect.width
+        first = triggerRect.left < overflowX / 2
+        last =
+          window.innerWidth - triggerRect.left - triggerRect.width <
+          overflowX / 2
+        break
+      case PopoverPosition.ToTheLeft:
+      case PopoverPosition.ToTheRight:
+        // When the dialog is left or right of the trigger
+        // First: top edge
+        // Last: bottom edge
+        const overflowY = dialogRect.height - triggerRect.height
+        first = triggerRect.top < overflowY / 2
+        last =
+          window.innerHeight - triggerRect.top - triggerRect.height <
+          overflowY / 2
+        break
+      default:
+        break
+    }
+
+    return {
+      first,
+      last,
+    }
+  }
+
+  private handleTriggerClick = (): void => {
+    const {showEvent} = this.props
+
+    if (showEvent === PopoverInteraction.Click) {
+      this.handleShowDialog()
+    }
+  }
+
+  private handleTriggerMouseOver = (): void => {
+    const {showEvent} = this.props
+
+    if (showEvent === PopoverInteraction.Hover) {
+      this.handleShowDialog()
+    }
+  }
+
+  private handleTriggerMouseOut = (): void => {
+    const {hideEvent} = this.props
+
+    if (hideEvent === PopoverInteraction.Hover) {
+      this.handleHideDialog()
+    }
+  }
+
+  private handleClickOutside = (): void => {
+    const {hideEvent} = this.props
+
+    if (hideEvent === PopoverInteraction.Click) {
+      this.handleHideDialog()
+    }
+  }
+
+  private handleUpdateTriggerRect = (): void => {
+    if (!this.triggerRef.current) {
+      return
+    }
+
+    this.setState({
+      triggerRect: this.triggerRef.current.getBoundingClientRect(),
+    })
+  }
+
+  private handleShowDialog = (): void => {
+    this.handleUpdateTriggerRect()
+
+    window.addEventListener('resize', this.handleUpdateTriggerRect)
+    window.addEventListener('scroll', this.handleUpdateTriggerRect)
```

Comment:
I think this will miss the case of the element moving on screen even if the window doesn't resize or the user doesn't scroll. For example if a sidebar is closed, everything on the side of it will shift. Kinda a minor concern that seems hard to fix.

Reply:
Ah that's a good point. I have no idea how do fix that unless there was some sort of fancy `componentDidMove` event